File: typescript/bob/bob.ts
class AnswerFactory {
    public static create(value: string): Sentence {
        if (this.isSilence(value)) {
            return new Silence()
        } else if (this.isShouting(value) && !this.isOnlyNumbers(value)) {
            return new Shouting()
        } else if (this.isQuestion(value)) {
            return new Question()
        }

        return new Sentence()
    }

    private static isShouting(value: String): boolean {
        return value === value.toUpperCase()
    }

    private static isOnlyNumbers(value: string): boolean {
        return /^([1-4],?\s?\??)+$/g.test(value)
    }

    private static isQuestion(value: string): boolean {
        return value.endsWith('?')
    }

    private static isSilence(value: string): boolean {
        return value.trim() === ''
    }
}

class Sentence {
    answer: string = 'Whatever.'
}

class Shouting implements Sentence {
    get answer(): string {
        return 'Whoa, chill out!'
    }
}

class Question implements Sentence {
    get answer(): string {
        return 'Sure.'
    }
}

class Silence implements Sentence {
    get answer(): string {
        return 'Fine. Be that way!'
    }
}

class Bob {
    hey(message: string): string {
        return AnswerFactory.create(message).answer
    }
}

export default Bob

File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import           Data.Char (isAlpha)
import           Data.Text (Text)
import qualified Data.Text as T

responseFor :: String -> String
responseFor = responseFor' . T.pack

responseFor' :: Text -> String
responseFor' stm
  | silence = "Fine. Be that way!"
  | isQuestion && isYelling = "Calm down, I know what I'm doing!"
  | isQuestion = "Sure."
  | isYelling = "Whoa, chill out!"
  | otherwise = "Whatever."
  where
    silence = T.null . T.strip $ stm
    isQuestion = (T.last . T.strip $ stm) == '?'
    isYelling = hasAlpha && stm == T.toUpper stm
    hasAlpha = T.any isAlpha stm
